### ISSUE TITLE
[1.10.1] station nil reference in ProcessRequest

### ISSUE THREAD

**Issue by Xyic0re** [Yousei9/Logistic-Train-Network#157]
factorio -: 0.17.6 build 43339
ltn -: 1.10.1
ltn tracker -: 0.9.1
ltn content reader -: 0.2.1
--
2x depot(s)
2x provider stations (mines); 1 stone , 1 iron
2x requester stations (1 stone, 1 iron)

depot = empty > both trains in service > requesters requesting deliveries > 1 train arrives back to depot > game crash

almost immediately after the train arrives in the depot.

------------------ 

38699.070 Error MainLoop.cpp:1129: Exception at tick 12574271: The mod LTN - Logistic Train Network caused a non-recoverable error.
Please report this error to the mod author.

Error while running event LogisticTrainNetwork::on_tick (ID 0)
__LogisticTrainNetwork__/control.lua:1515: attempt to index field 'station' (a nil value)
stack traceback:
	__LogisticTrainNetwork__/control.lua:1515: in function 'ProcessRequest'
	__LogisticTrainNetwork__/control.lua:1075: in function <__LogisticTrainNetwork__/control.lua:987>
38706.345 Quitting: user-quit.
38706.480 DSound: Stopping voice
38706.480 DSound: Joining thread
38706.482 DSound: Exit _dsound_update; tid=14292
38706.482 DSound: Waiting for voice to stop ... signaled
38706.482 DSound: Joined thread
38706.482 DSound: Destroying thread
38706.482 DSound: Thread destroyed
38706.482 DSound: Releasing buffer
38706.482 DSound: Voice stopped
38706.482 DSound: Deallocating voice
38706.482 DSound: Deallocated voice
38706.528 Steam API shutdown.
38706.569 Goodbye

**Comment by Yousei9** [comment]
Rules for bug reporting apply to github as well.
https://forums.factorio.com/viewtopic.php?f=214&t=60889

With the information provided I can do nothing except wrapping pointless checks tanking performance around every call to any function.

**Comment by Xyic0re** [comment]
I would like to help, what other information can I provide?

**Comment by Xyic0re** [comment]
link to the save file - https://drive.google.com/open?id=1KCdWDWcCJX9MeSGNTzhdTUtekGJyjEEJ

**Comment by Xyic0re** [comment]
it seems like the problem is with the stone mine. I will try to reset it and see what happens. I will have to do this in about 10 hours or so.

**Comment by Yousei9** [comment]
I see now, that's some impressive timing going on.

The train Garbfink arrives with some left over iron ore. Your circuitry disables the stop causing the train to repath to the other Depot. Somehow keeping train.station to the disabled depot valid.

Then the following happens within the same tick: 
LTN checks for available trains, including a check if the train.station exists. Between this and where LTN tries to set the lamp signal the train has left the station enough for train.station no longer exists.

I would have never thought it possible for train.station to become nil within the same tick.

**Comment by Yousei9** [comment]
Fixed for next release.

Your circuitry makes trains with residue ping pong between depots. Nonsensical but kinda amusing.

**Comment by Xyic0re** [comment]
Thanks a lot for your help, I appreciate it. Implemented the depot fix and seems to have resolved the issue.
What my circuitry tries to do is this;
defines 2 conditions; 1: set the minimum amount of item to have in storage 2: set a percentage of the maximum storage to fill up to.
easier to explain with a scenario; stone requester has max capacity of 100K stone.
You set the minimum to have in storage to 10K
You set the fill to percentage to 90%
when the stone in storage drops below 10K the SR latch will latch.
This will send the amount needed to fill the storage to your desired number to the ltn network
if the condition on the requester is green 1, it will send the little blue requester signal too.
the trains will come, when the storage reaches 90% (90k) full the SR latch will unlatch, and the amount missing will no longer be sent to the ltn network.

basically I want the storage to use what it has, until the minimum limit, and then try to fill to what I have requested, or to the percentage I have set.

In the current map it is ping ponging a lot because the stone mine has next to no ore, since I only just set it up and didn't wait for it to fill before setting up a requester.

Does this make sense? sorry if its not a great description.

**Comment by Yousei9** [comment]
LTN could do that for you.
Normally you'd set requested amount to 90% storage capacity and request threshold to one full train. That makes LTN keep it stocked by only sending full train loads maximizing train utilization.

If for whatever reason you want a thundering train horde once a requester drops below say 50% as in the example you gave.
set requested amount to 90% storage capacity and request threshold to 50% capacity.
